## js

### Promise.all()

Promise.all()接受一个iterator

```js
       function pAll(_promises) {
            return new Promise((resolve, reject) => {
                const promises = Array.from(_promises);//转换成数组
                const r = [];//promise结果
                const len = promises.length;//长度
                let count = 0;//成功数量
                for (let i = 0; i < len; i++) {
                    Promise.resolve(promises[i])//传入的参数不一定是promise,通过Promise.resolve转换成promise
                        .then((o) => {
                            r[i] = o;
                            if (++count === len) {
                                resolve(r);
                            }
                        })
                        .catch((e) => {
                            reject(e);//有一个错误就表示失败
                        });
                }
            });
        }
        debugger
        pAll([1, 2, 3]).then(res => console.log(res))
        pAll([1, 2, Promise.resolve(1)]).then(res => console.log(res))
        pAll([1, Promise.reject(2), Promise.resolve(3)]).then(res => console.log(res))
```



### Promise

```js
class QWEPromise {
	constructor(executor) {
		this.status = PROMISE_STATUS_PENDING;
		this.value = undefined;
		this.reason = undefined;
		this.onFulfilledFns = [];
		this.onRejectedFns = [];

		const resolve = (value) => {
			if (this.status === PROMISE_STATUS_PENDING) {
				// 添加微任务
				queueMicrotask(() => {
					if (this.status !== PROMISE_STATUS_PENDING) return;
					this.status = PROMISE_STATUS_FULFILLED;
					this.value = value;
					this.onFulfilledFns.forEach((fn) => {
						fn(this.value);
					});
				});
			}
		};

		const reject = (reason) => {
			if (this.status !== PROMISE_STATUS_PENDING) return;

			queueMicrotask(() => {
				if (this.status !== PROMISE_STATUS_PENDING) return;
				this.reason = PROMISE_STATUS_REJECTED;
				this.reason = reason;
				// this.onRejected(this.reason);
				this.onRejectedFns.forEach((fn) => {
					fn(this.reason);
				});
				console.log("reject被调用了");
			});
		};

		try {
			executor(resolve, reject);
		} catch (err) {
			reject(err);
		}
	}
	then(onFulfilled, onRejected) {
		return new QWEPromise((resolve, reject) => {
			//在then调用之前，状态已经确定
			console.log("dads", this.status);
			if (this.status === PROMISE_STATUS_FULFILLED && onFulfilled) {
				// try {
				// 	const value = onFulfilled(this.value);
				// 	resolve(value);
				// } catch (err) {
				// 	reject(err);
				// }
				execFunctionWithCatchError(onFulfilled, this.value, resolve, reject);
			}
			if (this.status === PROMISE_STATUS_REJECTED && onRejected) {
				console.log("sffs222");
				// try {
				// 	const reason = onRejected(this.reason);
				// 	resolve(reason);
				// } catch (err) {
				// 	reject(err);
				// }
				execFunctionWithCatchError(onRejected, this.reason, resolve, reject);
			}
			if (this.status === PROMISE_STATUS_PENDING) {
				console.log("sffs");
				this.onFulfilledFns.push(() => {
					// try {
					// 	const value = onFulfilled(this.value);
					// 	resolve(value);
					// } catch (err) {
					// 	reject(err);
					// }
					execFunctionWithCatchError(onFulfilled, this.value, resolve, reject);
				});
				this.onRejectedFns.push(() => {
					// try {
					// 	const reason = onRejected(this.reason);
					// 	resolve(reason);
					// } catch (err) {
					// 	reject(err);
					// }
					execFunctionWithCatchError(onRejected, this.reason, resolve, reject);
				});
			}
		});
	}
}

const promise = new QWEPromise((resolve, reject) => {
	console.log("状态pending");
	resolve(1111);
});
// promise.then((res) => {
// 	console.log("res", res);
// });
// setTimeout(() => {
// 	promise.then(
// 		(res) => {
// 			console.log("setTimeout", res);
// 		},
// 		(err) => {
// 			console.log("err", err);
// 		}
// 	);
// });
promise
	.then((res) => {
		console.log("res1", res);
		return "aaa";
	})
	.then((res) => {
		console.log("res2", res);
	});
```



```js
class MyPromise {
	constructor(executor) {
		this.initValue();
		this.initBind();
		try {
			executor(this.resolve, this.reject);
		} catch (e) {
			this.reject(e);
		}
	}
	initValue() {
		// 初始化值
		this.PromiseResult = null; // 终值
		this.PromiseState = "pending"; // 状态
		this.onFulfilledCallbacks = []; // 保存成功回调
		this.onRejectedCallbacks = []; // 保存失败回调
	}
	initBind() {
		this.resolve = this.resolve.bind(this);
		this.reject = this.reject.bind(this);
	}
	resolve(value) {
		// state是不可变的
		if (this.PromiseState !== "pending") return;
		// 如果执行resolve，状态变为fulfilled
		this.PromiseState = "fulfilled";
		// 终值为传进来的值
		this.PromiseResult = value;
		// 执行保存的成功回调
		while (this.onFulfilledCallbacks.length) {
			this.onFulfilledCallbacks.shift()(this.PromiseResult);
		}
	}

	reject(reason) {
		// state是不可变的
		if (this.PromiseState !== "pending") return;
		// 如果执行reject，状态变为rejected
		this.PromiseState = "rejected";
		// 终值为传进来的reason
		this.PromiseResult = reason;
		// 执行保存的失败回调
		while (this.onRejectedCallbacks.length) {
			this.onRejectedCallbacks.shift()(this.PromiseResult);
		}
	}

	then(onFulfilled, onRejected) {
		// 接收两个回调 onFulfilled, onRejected

		// 参数校验，确保一定是函数
		onFulfilled =
			typeof onFulfilled === "function" ? onFulfilled : (val) => val;
		onRejected =
			typeof onRejected === "function"
				? onRejected
				: (reason) => {
						throw reason;
				  };

		var thenPromise = new MyPromise((resolve, reject) => {
			const resolvePromise = (cb) => {
				try {
					const x = cb(this.PromiseResult);
					if (x === thenPromise) {
						// 不能返回自身哦
						throw new Error("不能返回自身。。。");
					}
					if (x instanceof MyPromise) {
						// 如果返回值是Promise
						// 如果返回值是promise对象，返回值为成功，新promise就是成功
						// 如果返回值是promise对象，返回值为失败，新promise就是失败
						// 谁知道返回的promise是失败成功？只有then知道
						x.then(resolve, reject);
					} else {
						// 非Promise就直接成功
						resolve(x);
					}
				} catch (err) {
					// 处理报错
					reject(err);
				}
			};

			if (this.PromiseState === "fulfilled") {
				// 如果当前为成功状态，执行第一个回调
				resolvePromise(onFulfilled);
			} else if (this.PromiseState === "rejected") {
				// 如果当前为失败状态，执行第二个回调
				resolvePromise(onRejected);
			} else if (this.PromiseState === "pending") {
				// 如果状态为待定状态，暂时保存两个回调
				this.onFulfilledCallbacks.push(resolvePromise.bind(this, onFulfilled));
				this.onRejectedCallbacks.push(resolvePromise.bind(this, onRejected));
			}
		});

		// 返回这个包装的Promise
		return thenPromise;
	}
}
const test3 = new MyPromise((resolve, reject) => {
	resolve(100); // 输出 状态：成功 值： 200
	// reject(100) // 输出 状态：失败 值：300
})
	.then(
		(res) => 2 * res,
		(err) => 3 * err
	)
	.then(
		(res) => console.log(res),
		(err) => console.log(err)
	);
console.log(test3);
```

### Iterator

Iterator时一个特殊的对象：

- 具有next()方法，调用该方法会返回一个结果对象。
- 结果对象有两个属性：value 和 done。
- value代表具体的值；done表示遍历是否完成。
- 内部有一个指针，指向数据结构的起始位置。每调用一次next()方法，指针都会向后移动一个位置，知道指向最后一个位置

```js
function createIterator(arr) {
	let i = 0;
	return {
		next: () => {
			let done = i >= arr.length;
			let value = !done ? arr[i++] : undefined;
			return {
				value,
				done,
			};
		},
	};
}
let iterator = new createIterator([1, 2, 3]);
console.log(iterator.next()); //{ value: 1, done: false }
console.log(iterator.next()); //{ value: 2, done: false }
console.log(iterator.next()); //{ value: 3, done: false }
console.log(iterator.next()); //{ value: undefined, done: true }
```

> ES6 规定，默认的 Iterator 接口部署在数据结构的Symbol.iterator属性，或者说，一个数据结构只要具有Symbol.iterator属性，就可以认为是“可遍历的”（iterable）。

给Iterator添加Symbol.iterable属性，使其可被遍历。

```js
let arr = [1, 2, 3];
arr[Symbol.iterator] = function () { //不能用箭头函数
	let i = 0;
	return {
		next: () => {
			let done = i >= this.length;
			let value = !done ? this[i++] : undefined;
			return { value, done };
		},
	};
};
for (let item in arr) {
	console.log(item);
}
let iteratorObj = arr[Symbol.iterator]();

console.log(iteratorObj.next()); //{ value: 1, done: false }
console.log(iteratorObj.next()); //{ value: 2, done: false }
console.log(iteratorObj.next()); //{ value: 3, done: false }
console.log(iteratorObj.next()); //{ value: undefined, done: true }
```



